proposals/xxxx-experimental-dxil.md

```diff
@@ -0,0 +1,195 @@
+ * Long term experiments (and potentially extensions) aren't currently feasible
+ * Focused on core api system (hlsl instrinsics and DXIL ops)
+ * Works within the current intrinsics/DXIL op mechanisms
+ * Minimizes overall changes to the system and IHV processes
+ * Straightforward transition route from experimental to stable
+ * IHV drivers can support both experimental and stable versions of an op simultaneously
+   * simplifies migrations
+
+## Non-goals
+Future proposals may address the topics below but this proposal seeks to be a
+smaller isloated change. It intends to solves immediate term challenges
+without investing significant engineering efforts into a generalized solution.
+That said, an attempt is made to avoid proposals that preclude a generalized
+solution. Thus this proposal explicitly avoids addressing these issues:
+ * Full scale generalized extension system
+ * Process development to enable asynchronous non-colliding development
+ * Metadata/RDAT/PSV0/Custom lowering are out of scope for this document
+
+
+
+## Potential DXIL Op Solutions
+
+### Top 1 bit as "is experimental" flag
+
+The top bit of all opcodes is a flag stating if the opcode is experimental.
+
+No structural or shape changes to the DXIL occur, simply the fact that the opcode
+has the high bit set informs that it is experimental. This makes it very easy
+for the compiler and drivers to detect experimental opcodes. When an opcode is
+transistioned to stable the opcode needs to be assigned a stable number.
+This splits the 4 billion opcode space into two 2 billion partions. One for
+stable one for experimental.
+
+This is the simpliest proposal with the least invasive set of changes.
+
+Pros:
+ * Very simple
+ * Quick to implement
+ * Could be implemented "by hand" today by hard coding opcodes
+Cons:
+ * Not a solution for extensions
+ * transistion from experimental to stable isn't just unsetting the bit
+   * other stable ops may have already taken that number
+   * complicates the experimental->stable mapping
+
+### Top 8 bits as "opcode partition" value
+This is pretty much identical to the 1 bit flag proposal except there are 256
+partitions with 16 million opcodes each. The key difference is that it unlocks
+extension potential as extension developers such as IHVs could reserve a
+partition for their own use without collision with other opcodes.
+
+| Partition | Use |
+|-----------|-----|
+| 0 | stable |
+| 1 | experimental |
+| 2 | extension foo |
+| .. | extension .. |
+| 255 | extension 255 |
+
+
+Pros:
+ * Fairly simple
+ * Quick to implement
+ * Enables basic opcodes extension system
+Cons:
+ * transistion from experimental to stable isn't just clearing the partition
+   * other stable ops may have already taken that number
+   * complicates the experimental->stable mapping
+
+### Top 16 bits as "opcode partition" value
+Identical concept as above but with 64k partitions, each with 64k opcodes.
```

Comment:
This would be my preferred solution. I can't see any reason to have more than 64k core DXIL ops. In any case, we could always extend these with another table entry if necessary.

This gives you more space for the feature separation I mentioned earlier.

Also, more space for table/partition ID makes it much easier to partition that space into several sub-spaces, which could be a useful expansion in the future (like: Core/Experimental, Official Extensions, IHV Extensions, Private Extensions). This would keep IR and code compatibility between all compiler branches and avoid synchronization and disruption across organizations for shifting table (partition) IDs.

Which is why I think we should partition this with 16-bits for table (partition) ID and 16-bits for DXIL OpCode.